cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -41,7 +49,11 @@ class Node0Impl extends TIRDataFlowNode0 {
   DataFlowType getType() { none() } // overridden in subclasses
 
   /** Gets the instruction corresponding to this node, if any. */
-  Instruction asInstruction() { result = this.(InstructionNode0).getInstruction() }
+  Instruction asInstruction() {
+    result = this.(InstructionNode0).getInstruction()
+    or
+    result = Ssa::getIRRepresentationOfOperand(this.(SingleUseOperandNode0).getOperand())
```

Comment:
There's probably some room for refactoring here by creating an abstract `InstructionNode0` with two implementations: One that extends `TInstructionNode0`, and one that extends `SingleUseOperandNode0`. Then `asInstruction()` could be a short and sweet one-liner just like `asOperand`. Such a refactoring would also mean we hopefully wouldn't have to always consider these two cases separately whenever we interact with `InstructionNode0`.

Reply:
Done. This makes it look a whole lot nicer actually.

Reply:
Awesome. Yeah, I think all the added complexity is pretty much invisible to `DataFlowUtil.qll` now 👍.